### ISSUE TITLE
Submodules for vendor directory?

### ISSUE THREAD

**Issue by JohnMaguire** [govend/govend#41]
Thanks for govend. It seems to do what the others (Godep, gopm) cannot: put the sqlite3 driver (*including* the .c code!) into `vendor/`.

However, when I try to call `govend` to add a new dependency, I get a username/password prompt for Github.

```
-> % govend github.com/moiron/sqlx
Username for 'https://github.com': 
```

Is there a way to get it to default to doing the clone with SSH (i.e. `git clone git@github.com:moiron/sqlx.git`)?

*EDIT*: Just realized I had a typo in the URL which is the root cause of my issue... (Github assumes it is a private repository.)

Also, I noticed you commit your `vendor/` directory instead of adding it to `.gitignore` and calling `govend` locally. I was curious if there is a reason to do this?

**Comment by JohnMaguire** [comment]
Sorry, just saw #35 and will discuss the first part of this there.

However, as for the second part of the above, I realized it would be in order for `go get` to use the dependencies out of it. Have you considered using git submodules for the `vendor/` directory by leaving the `.git` directory intact?

**Comment by jackspirou** [comment]
Hi @JohnMaguire 

The thought of using git submodules inside of `vendor/` has crossed my mind, but there are a few issues that worried me with such an approach:

* git submodules add complexity to the code 
* having a `vendor.yml` or some manifest file accomplishes the only need we have of a git submodule, which is to have a record of the current revision version
* not all dependency repositories are git (some are svn, bzr) and so how should we treat those? the `vendor.yml` or manifest allows all repos from any VCS host to be handled in the same way
* for users that want to gitignore `vendor/`, the only way is to record the rev versions is via some manifest  `vendor.yml` file outside of `vendor/`.  This feature again eliminates the need for git submodules 
* git submodules have a bad rep (not a technical reason, but still worth mentioning since others execute commands on them that might mess up `govend`)

I know I just listed a bunch of concerns and disadvantages, but I am very open to hearing why you think they would be beneficial and give us something thats lacking, or do something better than its done today.

**Comment by JohnMaguire** [comment]
Hey Jack,

I can appreciate those concerns, and agree with you that it would add complexity (not just to the code, but also for users who are unfamiliar with submodules — as you said, they aren’t technically bad, but people struggle when cutting their teeth on them.)

I don’t think that the use of sub-modules conflicts with the “vendor.yml”. The file is great for developers to specify commits in to create the “vendor/“ directory with. Specifically, it should be run by build tools. At the same time, you may want to check out a development branch of the submodule in the “vendor/“ directory to test an upstream change with. By using submodules in “vendor/“ this would be as easy as adding a remote and checking out your own branch (like you would do with a standard Golang package).

While “vendor.yml" eliminates the need for submodules in terms of locking to a specific commit, it still provides use for those wish to gitignore “vendor/“, like you said. This is great, but if I gitignore “vendor/“, using `go get github.com/govend/govend` <http://github.com/govend/govend%60> will give me my `$GOPATH` dependencies, when I really wanted the dependencies listed in `vendor.yml`. Since `go get` doesn’t support reading the YML file, having submodules is a great alternative.

Not using submodules works too (the approach you seem to have taken here), but I consider it dirty (perhaps only ideologically):

- Committing code that is unrelated to my project
- Allows for possibility of accidentally editing and committing code in “vendor/“
- Pollutes git log and stats, and overall size of the repository (and since all your depends likely have more churn than your actual project code if you update them, the noise ratio is very high)
- There are also license violation concerns, especially in regards to the GPL (i.e. you may not ship GPL-licensed code in a non-GPL licensed project)

John

**Comment by JohnMaguire** [comment]
Hey Jack,

I can appreciate those concerns, and agree with you that it would add complexity (not just to the code, but also for users who are unfamiliar with submodules — as you said, they aren’t technically bad, but people struggle when cutting their teeth on them.)

I don’t think that the use of sub-modules conflicts with the “vendor.yml”. The file is great for developers to specify commits in to create the “vendor/“ directory with. Specifically, it should be run by build tools. At the same time, you may want to check out a development branch of the submodule in the “vendor/“ directory to test an upstream change with. By using submodules in “vendor/“ this would be as easy as adding a remote and checking out your own branch (like you would do with a standard Golang package).

While “vendor.yml" eliminates the need for submodules in terms of locking to a specific commit, it still provides use for those wish to gitignore “vendor/“, like you said. This is great, but if I gitignore “vendor/“, using `go get github.com/govend/govend` will give me my `$GOPATH` dependencies, when I really wanted the dependencies listed in `vendor.yml`. Since `go get` doesn’t support reading the YML file, having submodules is a great alternative.

Not using submodules works too (the approach you seem to have taken here), but I consider it dirty (perhaps only ideologically):

* Committing code that is unrelated to my project
* Allows for possibility of accidentally editing and committing code in “vendor/“
* Pollutes git log and stats, and overall size of the repository (and since all your depends likely have more churn than your actual project code if you update them, the noise ratio is very high)
* There are also license violation concerns, especially in regards to the GPL (i.e. you may not ship GPL-licensed code in a non-GPL licensed project)

John

**Comment by jackspirou** [comment]
@JohnMaguire i think this is turning out to be a great discussion.  Here are some more points/responses:

> I don’t think that the use of sub-modules conflicts with the “vendor.yml”. The file is great for developers to specify commits in to create the “vendor/“ directory with.

Agreed that the manifest file could still work well with git submodules, a very fair point.

> if I gitignore “vendor/“, using go get github.com/govend/govend will give me my $GOPATH dependencies, when I really wanted the dependencies listed in vendor.yml. Since go get doesn’t support reading the YML file, having submodules is a great alternative.

I believe it is the repo owners responsibility to include the `vendor/` directory if they want to be `go get`-able, that only seems reasonable (also we can't hold everyone's hand that much)

> [without git submodules it is possible to ] commit code that is unrelated to my project [and] allow for possibility of accidentally editing and committing code in “vendor/“

A valid point, but one already deliberated/decided by the Go Team when they decided on the `vendor/` directory, it was a lively discussion :stuck_out_tongue_winking_eye: 

> There are also license violation concerns, especially in regards to the GPL (i.e. you may not ship GPL-licensed code in a non-GPL licensed project)

So if you use git submodules instead of a normal directory structure this is legally no longer an issue?  I ask because I legitimately have no idea...

One point that wasn't address was the comment:
> not all dependency repositories are git (some are svn, bzr) and so how should we treat those? the vendor.yml or manifest allows all repos from any VCS host to be handled in the same way

Did you have any thoughts on deps that do not come from a git repo?  

Also what about the pruning process?  How does git submodules fit with that process?

Thanks for keeping this conversation going. Again I find it very helpful!

**Comment by JohnMaguire** [comment]
Sorry, it's been a few days, but a couple responses! :) You make some excellent points here. :+1: 

> I believe it is the repo owners responsibility to include the `vendor/` directory if they want their project to be `go get`-able, that only seems reasonable (also we can't hold everyone's hand that much)

I absolutely agree! And currently, you support this by creating a "vendor/" directory with dependencies checked out. My main two qualms with it are:

1. I prefer smaller repositories that contain only files relevant to them, with upstream vendors providing their own source. i.e. I should not be responsible for including upstream. It is quite possible that this is simply bias.

2. Cluttered Git history. Again, I think this is a mostly minor complaint, but I really don't like having "5 million lines" committed, that I didn't touch. It skews the true nature of what is happening within the repository. This is probably more important to me, specifically, as I am deeply into analytics, and think that repository analytics are a hugely untapped source of information. In fact, I have some ideas on how to capitalize on this.

An additional benefit of using submodules, in my opinion, is that it makes it easy to checkout different branches of the submodule (for example, to test something in a `dev` branch you pushed -- or to swap out the remote [upstream] for your own [fork]).

> So if you use git submodules instead of a normal directory structure, this is legally no longer an issue? I ask because I legitimately have no idea...

IANAL. I really didn't dig up information before making that claim, but I believe it to be true. This StackOverflow answer explains quite well the train of thought I was following when I made that claim: https://programmers.stackexchange.com/questions/240633/is-licensing-an-issue-for-git-submodules

`.gitmodules` is more like a dependencies list (think `requirements.txt`). It tells the user/system to go fetch some dependencies that aren't mine. Committing code under `vendor/` is integrating upstream's code into my project.

> not all dependency repositories are git (some are svn, bzr) and so how should we treat those? the vendor.yml or manifest allows all repos from any VCS host to be handled in the same way

I realized I forgot to address this one earlier. I do not have a great answer for this yet.

> Also what about the pruning process? How does git submodules fit with that process?

It really doesn't. I also don't think submodules necessarily need to be mandatory. I imagine it as a flag that could be stored in the manifest file, or triggered at the command line. This way you have two options:

1. Use submodules for development process
2. Trigger a flag to not use submodules (and prune) for build process

It is fairly standard for a development flow to not look exactly like the build flow. Simplicity is of course key to any tool, and maybe this would be unadvantageous (too many feature toggles will kill any project.)

**Comment by JohnMaguire** [comment]
Interestingly, the link you posted makes a few points that I think using submodules solves (ones I brought up):

> 2 It discourages contribution to upstream (because it is quicker to fix
>   it in the vendored copy, right?)

With submodules, you can quickly add your own fork's remote URI, test it, and then merge back upstream.

> 3 As a consequence it creates diverging versions of the same library

See above.

> 4 And possibly licensing issues (as observed with ruby)

A point I addressed.

> 5 It puts a lot of burden unto package maintainers, as distributions
>   normally disallow vendoring (for security- and maintability concerns)

This is part of why I like submodules. It doesn't completely solve this issue, but updating a submodule doesn't require use of special tools, like `govend`. It also is still pointing to the canonical source, which is pretty useful. I know the manifest file does this too but... well, I don't really have a great argument here.

**Comment by jackspirou** [comment]
@JohnMaguire thanks for the thoughtful responses.  I just got around to reading them.

I think you have made some interesting and valid points, and I need to mull it over in my brain.  Especially the following comment: 

> * Use submodules for development process
* Trigger a flag to not use submodules (and prune) for build process

Introducing submodules would be an enormous effort that could complicate the codebase further, but i am considering some of the benefits we discussed with a flag option.  This probably not something im going to be spending time on developing at this point, but if you feel passionate I would consider a PR or starting an experimental branch.
